GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -156,10 +151,10 @@ void InitializeComponent()
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2002:DoNotLockOnObjectsWithWeakIdentity", Justification = "It looks like such lock was made intentionally but it is better to rewrite this")]
         protected override void Dispose(bool disposing)
         {
-            _shouldRun = false;
-
             if (disposing)
             {
+                _shouldRun = false;
+                JoinBackgroundThreadAndPendingAsyncOperations();
```

Comment:
So I think you're onto something here (it's touching an area I've been thinking for a while could be improved), but I don't believe this is the correct way to do it. Git Extensions is challenging (compared to e.g. a WPF app) since there so many forms that get opened and closed, with different operations needing to be scoped to different forms. My preference for a high-level approach would be:

1. Make sure a `CancellationToken` is associated with each form, and gets cancelled when the form is set to close (after the last point where closing the form can be cancelled).
1. Make sure a `JoinableTaskCollection` is associated with each form.
1. Make sure background operations associated with the form are tracked by the `JoinableTaskCollection` belonging to that form. These background operations should list for the cancellation token, and cancel any unnecessary work when the form is closing.
1. Join the JoinableTaskCollection when the form is closing, which I believe would mean the Dispose method.

We may end up with some cases where forms get delayed in closing, but we can resolve these by ensuring that background operations correctly use the cancellation token associated with the form.

Reply:
This sounds very reasonable.
This background thread should be eliminated in a follow-up PR. (Then the ugly wait loops will not be needed anymore.)
Here, I have just added thread joining and factored out duplicate code.

Reply:
merged `sharwell/rm-background-thread`